### PR TITLE
Add 4 additional commands for build

### PR DESCRIPTION
**Pull request by xiadu94**
Added 4 CLI command regarding Issue [145](https://github.com/AzureCR/azure-cli/issues/145)

`acr build-task list-argument`
`acr build-task list-source-properties`
`acr build-task update-build`
`acr build-task cancel`

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 1]
```diff
@@ -251,6 +280,17 @@ def acr_build_task_update(cmd,  # pylint: disable=too-many-locals
 
     return build_task
 
+def acr_build_task_update_build(cmd,  # pylint: disable=too-many-locals
+                                client,
+                                build_id,
+                                registry_name,
+                                is_archive_enabled=None,
+                                resource_group_name=None):
+    if is_archive_enabled:
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Does this work if `is_archive_enabled` is set to `False`?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 2]
```diff
@@ -251,6 +280,17 @@ def acr_build_task_update(cmd,  # pylint: disable=too-many-locals
 
     return build_task
 
+def acr_build_task_update_build(cmd,  # pylint: disable=too-many-locals
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Is disabling pylint needed here?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py [hunk 3]
```diff
@@ -151,3 +155,4 @@ def load_command_table(self, _):  # pylint: disable=too-many-statements
                   table_transformer=build_output_format)
         g.command('logs', 'acr_build_task_logs', client_factory=cf_acr_builds,
                   table_transformer=None)
+        g.command('cancel', 'acr_build_task_cancel', client_factory=cf_acr_builds)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: nit: wasn't sure if you want to call this as `cancel-build` (and align with `update-build` or `list-builds`) otherwise the cancel seems to be on the build task.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py [hunk 4]
```diff
@@ -141,8 +142,11 @@ def load_command_table(self, _):  # pylint: disable=too-many-statements
         g.command('create', 'acr_build_task_create')
         g.command('show', 'acr_build_task_show')
         g.command('list', 'acr_build_task_list', table_transformer=build_task_output_format)
+        g.command('list-arguments', 'acr_build_arguments_list', client_factory=cf_acr_build_steps)
+        g.command('list-source-properties', 'acr_build_source_repository_properties_list')
         g.command('delete', 'acr_build_task_delete')
         g.command('update', 'acr_build_task_update')
+        g.command('update-build', 'acr_build_task_update_build', client_factory=cf_acr_builds)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: nit: group this with `list-builds`, `show-build`, etc.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py [hunk 5]
```diff
@@ -141,8 +142,11 @@ def load_command_table(self, _):  # pylint: disable=too-many-statements
         g.command('create', 'acr_build_task_create')
         g.command('show', 'acr_build_task_show')
         g.command('list', 'acr_build_task_list', table_transformer=build_task_output_format)
+        g.command('list-arguments', 'acr_build_arguments_list', client_factory=cf_acr_build_steps)
+        g.command('list-source-properties', 'acr_build_source_repository_properties_list')
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: IMO: `show-build-arguments` and `show-source-repository`.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 6]
```diff
@@ -149,6 +149,35 @@ def acr_build_task_list(cmd,
         cmd.cli_ctx, registry_name, resource_group_name, BUILD_TASKS_NOT_SUPPORTED)
     return client.list(resource_group_name, registry_name)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Two empty lines between methods.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 7]
```diff
@@ -149,6 +149,35 @@ def acr_build_task_list(cmd,
...
+                                                build_task_name,
+                                                resource_group_name=None):
+    _, resource_group_name = validate_managed_registry(
+        cmd.cli_ctx, registry_name, resource_group_name, BUILD_TASKS_NOT_SUPPORTED)
+    return client.list_source_repository_properties(resource_group_name, registry_name, build_task_name)
+
+def acr_build_task_cancel(cmd,
+                          client,
+                          registry_name,
+                          build_id,
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Do we want to support cancel last created build (in a build task) style? similar to logs.
/cc @ankurkhemani @northtyphoon @ehotinger

**Review by djyou — CHANGES_REQUESTED** [review]
Please also update tests, history, and version (if needed) once you finalize the commands but before merging.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py (hunk 1 above)
Comment (xiadu94) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Changed to `if is_archive_enabled is not None`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py (hunk 2 above)
Comment (xiadu94) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Removed.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py (hunk 3 above)
Comment (xiadu94) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: Kept the name `cancel` while removed the `client-factory` here.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py (hunk 4 above)
Comment (xiadu94) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: Changed.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py (hunk 5 above)
Comment (xiadu94) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: Changed.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py (hunk 6 above)
Comment (xiadu94) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Removed.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py [hunk 8]
```diff
@@ -141,13 +142,17 @@ def load_command_table(self, _):  # pylint: disable=too-many-statements
...
         g.command('delete', 'acr_build_task_delete')
         g.command('update', 'acr_build_task_update')
+        g.command('cancel', 'acr_build_task_cancel')
         g.command('run', 'acr_build_task_run', client_factory=cf_acr_builds,
                   table_transformer=build_output_format)
         g.command('list-builds', 'acr_build_task_list_builds', client_factory=cf_acr_builds,
                   table_transformer=build_output_format)
         g.command('show-build', 'acr_build_task_show_build', client_factory=cf_acr_builds,
                   table_transformer=build_output_format)
+        g.command('update-build', 'acr_build_task_update_build', client_factory=cf_acr_builds)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/commands.py]: Add `table_transformer=build_output_format`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py (hunk 7 above)
Comment (ehotinger) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: well cancellation is still not implemented AFAIK, unless someone is actively implementing it. So no cancellation should be in the CLI yet

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 9]
```diff
@@ -251,6 +279,17 @@ def acr_build_task_update(cmd,  # pylint: disable=too-many-locals
...
+                                client,
+                                build_id,
+                                registry_name,
+                                is_archive_enabled=None,
+                                resource_group_name=None):
+    if is_archive_enabled is not None:
+        return client.update(resource_group_name=resource_group_name,
+                             registry_name=registry_name,
+                             build_id=build_id,
+                             is_archive_enabled=is_archive_enabled)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: I think you can just return directly without checking if `is_archive_enabled` is None or not, otherwise this command doesn't print anything if `is_archive_enabled` wasn't specified. Also add `is_archive_enabled` to `_params.py` with description if you haven't. Were you able to run this command and get expected results from server?

IMO, also rename this argument to `archive_enabled` or even `no_archive` (to be consistent with `no_push` or `no_logs`), but I would wait to see what others might say.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 10]
```diff
@@ -149,6 +148,35 @@ def acr_build_task_list(cmd,
...
+    return client.list_source_repository_properties(resource_group_name, registry_name, build_task_name)
+
+def acr_build_task_cancel(cmd,
+                          client,
+                          registry_name,
+                          build_id,
+                          resource_group_name=None):
+    _, resource_group_name = validate_managed_registry(
+        cmd.cli_ctx, registry_name, resource_group_name, BUILD_TASKS_NOT_SUPPORTED)
+    return client.list(resource_group_name, registry_name, build_id)
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: cancel?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 11]
```diff
@@ -355,6 +355,24 @@
             az acr build-task list -r MyRegistry -o table
 """
 
+helps['acr build-task list-arguments'] = """
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: @djyou is this a common az cli pattern, instead of just returning normal json and --query for properties instead? I feel like it clutters the CLI significantly by first classing these commands

Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: Update the helps once we finalize the commands and parameters.

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: This is a different API that shows sensitive information (e.g., secrets in build args), which wasn't included in get build task path.

We can combine this with `acr build-task show` though, with a flag.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py (hunk 7 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: Let's remove it from this PR then and ship the other three commands.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 12]
```diff
@@ -403,6 +421,24 @@
             az acr build-task run -n MyBuildTask -r MyRegistry
 """
 
+helps['acr build-task update-build'] = """
+    type: command
+    short-summary: Patch the build properties.
+    examples:
+        - name: Update an existing build to be archived or not.
```
Comment (ehotinger) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: what does archiving a build mean? it also seems weird to make it a part of build-task, since they should be independent things, but i trust you guys have thought out all the weird cli/rp hierarchy stuff

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: I didn't totally agree, but the spirit was to put everything (except quick build) under build task.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py (hunk 6 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: No, not remove, we want 2 empty lines between methods.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 13]
```diff
@@ -119,26 +119,45 @@ def acr_build_task_show(cmd,
...
+                                                                  registry_name=registry_name,
+                                                                  build_task_name=build_task_name,
+                                                                  step_name=_get_build_step_name(build_task_name))
+        response['buildArguments'] = build_arguments
     except CloudError as e:
         if e.status_code != 404:
             raise
-        logger.warning("Could not get build task details. Build task basic information is printed.")
-
-    return build_task
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py]: I don't think you need to modify existing logic but before `return build_task` if `--secret` is specified, you also try to make the two extra requests and merge them to `build_task`.

You also need to handle permission issue gracefully here. Try assigning yourself as a reader on a registry and see what the error looks like, but in the code here we don't want to block the result but print a warning similar to the one we print on 404 error when build step couldn't be found.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 14]
```diff
@@ -134,6 +134,7 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('repository_url', options_list=['--context', '-c'], help="The full URL to the source code respository.")
         c.argument('commit_trigger_enabled', help="Indicates whether the source control commit trigger is enabled.", arg_type=get_three_state_flag())
         c.argument('git_access_token', help="The access token used to access the source control provider.")
+        c.argument('secret_details', options_list=['--secret'], help="Indicates whether the general information or secret details of builds should be returned.", action='store_true')
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Why not just call the var `secret` so you don't need to create a mapping using `options_list`

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 15]
```diff
@@ -341,9 +341,13 @@
     type: command
     short-summary: Get the properties of a specified build task.
     examples:
-        - name: Get the details of a build task, displaying the results in a table.
+        - name: Get the non-secret details of a build task, displaying the results in a table.
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py]: nit: keep the original string as is. Change the following one to "Get the details of a build task including secrets, displaying the results in a table".

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 16]
```diff
@@ -143,6 +144,7 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('top', help='Limit the number of latest builds in the results.')
         c.argument('build_id', help='The unique build identifier.')
         c.argument('build_status', help='The current status of build.', arg_type=get_enum_type(BuildStatus))
+        c.argument('no_archive', help='Indicates whether the image built should be pushed to the registry.', arg_type=get_three_state_flag())
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Update the help message.

Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py]: Please also rebase, add tests, and update history.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py [hunk 17]
```diff
@@ -383,6 +383,10 @@
         - name: Get the details of a build task, displaying the results in a table.
           text: >
             az acr build-task show -n MyBuildTask -r MyRegistry -o table
+
+        - name: Get the details of a build task including secrets.
```
Comment (sajayantony) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py:387]: >secrets [](start = 58, length = 7)

please rename this to --secure-properties
` az acr build-task show --with-secure-properties`
or ` az acr build show --secure-properties` if they only display secure proprties. 

@djyou?

**Review by sajayantony — CHANGES_REQUESTED** [review]
:clock1:

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands.py [hunk 18]
```diff
@@ -356,19 +356,54 @@ def test_acr_create_build_task(self, resource_group, resource_group_location):
                          self.check('[1].name', '{build_task_name2}')])
 
         # trigger a build from the build task
-        self.cmd('acr build-task run -n {build_task_name1} -r {registry_name} --no-logs',
+        response = self.cmd('acr build-task run -n {build_task_name1} -r {registry_name} --no-logs',
                  checks=[self.check('type', 'Microsoft.ContainerRegistry/registries/builds'),
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/test_acr_commands.py:360]: Fix the indentation here.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py [hunk 19]
```diff
@@ -158,6 +159,7 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('build_id', help='The unique build identifier.')
         c.argument('build_status', help='The current status of build.', arg_type=get_enum_type(BuildStatus))
         c.argument('image', arg_type=image_by_tag_or_digest_type)
+        c.argument('no_archive', help='Indicates whether the image built should be pushed to the registry.', arg_type=get_three_state_flag())
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py:162]: Fix the description.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py (hunk 17 above)
Comment (djyou) [reply src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_help.py:387]: `--with-secure-properties` sounds good, although it's a bit lengthy, how about `--with-secret`?

@xiadu94 Did you know if anyone else is using `--secret` or something similar?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/recordings/test_acr_create_replication.yaml [hunk 20]
```diff
@@ -1,27 +1,27 @@
 interactions:
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/tests/latest/recordings/test_acr_create_replication.yaml:1]: Could you revert the changes in other yaml files except the build_task one?

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 21]
```diff
@@ -252,6 +277,20 @@ def acr_build_task_update(cmd,  # pylint: disable=too-many-locals
     return build_task
 
 
+def acr_build_task_update_build(cmd,
+                                client,
+                                build_id,
+                                registry_name,
+                                no_archive,
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py:284]: Add a default value for `no_archive`, all properties are optional.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 22]
```diff
@@ -138,6 +139,30 @@ def acr_build_task_show(cmd,
...
+    try:
+        source_repository = client.list_source_repository_properties(resource_group_name,
+                                                                     registry_name,
+                                                                     build_task_name)
+        setattr(build_task, 'sourceRepository', source_repository)
+    except CloudError as e:
+        if e.status_code != 403:
+            raise
+        # pylint: disable=line-too-long
+        logger.warning("No permission to list source repository secret properties. Build task basic information is printed.")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py:154]: list -> get?

Also we are not printing basic information per se.

File: src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py [hunk 23]
```diff
@@ -138,6 +139,30 @@ def acr_build_task_show(cmd,
...
+    try:
+        build_arguments = client_build_steps.list_build_arguments(resource_group_name=resource_group_name,
+                                                                  registry_name=registry_name,
+                                                                  build_task_name=build_task_name,
+                                                                  step_name=_get_build_step_name(build_task_name))
+        setattr(getattr(build_task, 'properties'), 'buildArguments', list(build_arguments))
+    except CloudError as e:
+        if e.status_code != 403:
+            raise
+        logger.warning("No permission to list build arguments. Build task basic information is printed.")
```
Comment (djyou) [inline src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/build_task.py:165]: Same

**Review by djyou — COMMENTED** [review]
Just a few minor comments.
